View logs state

floffy-f opened the pull request:
# Visual indicators for presence of data in pages

Some minor changes are shown here, regarding the buttons of the *logs* page and the *registration* page.
The goal was for them to show some *visual hints* about the *content* of their related pages.

## Two buttons changed :

Initially, the two buttons went from `white` to `almostWhite` colors when clicked, as shown bellow :
![clean_logs_and_reg](https://user-images.githubusercontent.com/85450305/121659314-79171280-caa2-11eb-8247-80e01dc2cc18.png)

Now the registration button goes from `white`/`almostWhite` to `green` when the registered images are *ready* :
![green_reg_and_dark_logs](https://user-images.githubusercontent.com/85450305/121659356-8207e400-caa2-11eb-8240-45b7558447b8.png)

In addition to that, when logs appear the logs button's border goes from empty to dotted :
![dark_logs](https://user-images.githubusercontent.com/85450305/121659342-7e745d00-caa2-11eb-88b2-3a9b90bce88b.png)

It can also go to red `errorColor` when at least one error-log appears in the logs
![reg_green_and_logs_red](https://user-images.githubusercontent.com/85450305/121659319-7a483f80-caa2-11eb-9a61-302360f4eec2.png)


## Mechanisms :

### For the register button :

The `registeredImages` data-structure is a `Maybe (Pivot Image)`, so when the images have all gone through the _rank-reduction algorithm_, the `Model` field goes from `Nothing` to `Just (Pivot Image)`.

We use the `Maybe.Extra` package, which provides the function `isJust` that lets us know if the images are loaded in the `Registration` page or not.

### For the logs button

We have a new function `getMaxLevel` to know in a list of logs if one of them is an error :red_circle:.
We use this function to render the `headerBar` in each window, as an `Int` parameter.
It could be upgraded by using the adventages of *Elm* to make an adapted structure instead of an `Int`

mpizenberg commented:
Thanks for this PR! The code is working well but there are still brittle things and things that need changes so here is a detailed review.

In your imports, there is no need to expose `isJust` if you are always calling it qualified `Maybe.Extra.isJust`. In that case, you can simply write `import Maybe.Extra`.

Regarding visuals, there is a displacement of the text for the "Logs" tab when adding a border. Seeing that text not aligned with the rest is a bit unsettling. I'd suggest adding something like `Element.inFront` and putting a border on that thing, such that the original text below is not displaced. I also think using only the bottom border might be enough to grab the user attention. No need to add a full 360 border in my opinion.

Regarding the code, I think you are evaluating the `registrationState` and `logsState` too early in the call stack. This results in passing computed arguments to parts of functions that don't actually need them. In addition, it now seems that the function `pageHeaderElement` is trying to regroup view functions under the same function that should be different. The only two that are similar are the "Image" and "Config" tabs but it would be much appropriate now to just have a different function for each tab. Something like:

```elm
headerBar : List (Element Msg) -> Element Msg
headerBar pageTabs model =
    Element.row [ centerX, height (Element.px headerHeight) ]
        pageTabs
```

And those tabs would be computed directly by the caller, such as `viewLogs` which would be something like:

```elm
viewLogs : Model -> Element Msg
viewLogs ({ autoscroll, verbosity, logs, registeredImages } as model) =
    Element.column [ width fill, height fill ]
        [ headerBar
            [ imagesHeaderTab False
            , configHeaderTab False
            , registrationHeaderTab False registeredImages
            , logsHeaderTab True logs
            ]
        , ...
        ]
```

And as such, the `imagesHeaderTab` and `configHeaderTab` will not have to worry about logs or registration state. Of course, we can still have common helper functions or constants used in each header tab function if it makes sense.


------

For the detection of logs, I agree with your comment that `Int` is not the most appropriate type to describe if there exists logs. Also, I noted that the behavior implemented in this PR is different than the one suggested in issue #6. In the issue, it is suggested that the visual clue would indicate new logs. In this implementation, it actually indicates the presence of logs, whether they are new or not.

In your `getMaxLevel` function, you currently return 0 if there is an error, 1 if there is a warning, 2 if there is 0 or 1 log message that isn't a warning or error, and return 3 in the other cases. And you note that the limit between returning 2 or 3 is at 1 log because there is already one log done at the beginning of the application.

In my opinion, this approach will get potentially confusing for two reasons:

1. Using `Int` and returning 0 or 1 when there is an error or a warning strongly suggests that you are actually returning the lowest level of existing logs. But then the returned `2` is very suprising since it does not correspond to a level and does not correspond to a count either. Same for `3`. So it would indeed be better to return a custom type with the most appropriate names possible.
2. Choosing one message instead of zero as the limit for the visual indication is brittle and might break unexpectedly if anything else writes some logs at the image loading phase of the app.

So there ar two options to solve this. First option is the visual indication is for the presence of logs, and then it's better to be strict and keep the threshold at 0 log. In that case, you could keep and `Int` or rather a `Maybe Int` that would be the "lowest log level" if some exists or `Nothing` otherwise.

The second option is to indicate the presence of actual "new" logs. But in that case, it might require a bit more work since you'd need to differentiate "old" from "new" logs depending on if the user saw them already. So this might involve keeping the id of the last log that was "viewed" for each verbosity level. I don't think the added value is worth the time spent coding it for now, so I'd stick to the first option.

mpizenberg committed:
Change indicators in header bar

Visual indicator for logs apparition added :
* If no logs : no border
* If no error/warning logs : black dotted border
* If error/warning logs (at least one) : red dotted border
   /!\ -> behavior can be easily modified to handle a different
   color for
   warnings and error if needed
   Visual indicator for registration button too :
* If images shown : green background whatever the
   context
* If empty : normal white/almostWhite color

floffy-f committed:
[Visual indicator]
Little dots for notifications instead of borders.
Refactoring of the code of headerBar following the advices of Matthieu on the PR.

floffy-f committed:
[Visual indicator]
From borders to little dot + indicate new logs only

The borders became little dots at the top left of the button.
The frontiers between the buttons was added to avoid any doubt on the
position of the little dot.

The logs dot indicates only new logs, not the presence of any of those
anymore.

floffy-f committed:
[Visual indcations]
Format of logstype from Int to custom type for code readability

floffy-f commented:
# 3 last commits :

We use a little `svg` dot instead of the borders around the buttons.

The *logs dot* now retracts whenever the user watches the current logs.

The *dot for the `Registration` tab* does not disapear when the user clicks the navigation button.
It just stays around as long as results are being stored.

*Thin borders* were also added to the buttons, to separate them, but they are not mandatory by any mean, their only purpose is to avoid confusion about the position of the notification dots.
Feel free to remove them if they are not sexy enough : an idea could be to try to recenter the dots a bit to see if their position can feel a bit more obvious.


### Implementation :
 Two fields added to the model :
* `seenLogs` : the logs which were already there the last time the user went in the `Logs` tab.
* `notSeenLogs` : the new logs, unseen at the moment.

The previous `logs` field has disapeared.
It would correspond to the following equivalent expression :
```elm
logs : List { lvl : Int, content : String }
logs = List.concat [seenLogs, notSeenLogs]
```

### Visual results
Hence the following visuals :
![dots](https://user-images.githubusercontent.com/85450305/122065159-02577d80-cdf2-11eb-9516-4a35de5c016e.png)

### Quick refactoring

As suggested, the following functions have been implemented :
```elm
viewLogs : Model -> Element Msg
viewLogs ({ autoscroll, verbosity, logs, registeredImages } as model) =
    Element.column [ width fill, height fill ]
        [ headerBar
            [ imagesHeaderTab False
            , configHeaderTab False
            , registrationHeaderTab False registeredImages
            , logsHeaderTab True logs
            ]
        , ...
        ]
```
With :
```elm
imagesHeaderTab : Bool -> Element Msg

configHeaderTab : Bool -> Element Msg

registrationHeaderTab : Bool -> Maybe (Pivot Image) -> Element Msg

logsHeaderTab : Bool -> List { lvl : Int, content : String } -> Element Msg

headerBar : List (Element Msg) -> Element Msg
headerBar pageTabs =
    Element.row [ centerX, height (Element.px headerHeight) ]
        pageTabs
```

mpizenberg committed:
Expand formatting of all RunAlgorithm update branches

mpizenberg committed:
Factorize switchToLogsPage for the model

mpizenberg committed:
Reorder arguments of the goTo function

mpizenberg committed:
Use goTo function in runAndSwitchToLogsPage function

mpizenberg committed:
Fix swapped order between notSeenLogs and seenLogs

mpizenberg committed:
Directly use seenLogs in viewLogs since it was just concatenated

mpizenberg committed:
Remove unused code

mpizenberg committed:
Remove old commented code

mpizenberg committed:
Remove borders on header tabs

mpizenberg committed:
Remove unused old code

mpizenberg committed:
Move around misplaced comments after formatting

mpizenberg committed:
Factorize baseTabAttributes

mpizenberg committed:
Factorize littleDot

mpizenberg committed:
Dedup baseTabAttributes when adding notification dot

mpizenberg committed:
Extract buttonAttributes in let in

mpizenberg committed:
Remove need for maybe-extra dependency

mpizenberg reviewed (approved):
Thanks @floffy-f for the modifications, it's working very well and having those activity identifiers improves a lot the user experience in my opinion.

I've added many commits, mostly some code cleanup plus two changes:
1. As you suggested, the tabs borders are not really necessary anymore with the top right dot. I think that's clear enough.
2. I fixed the logs order. Indeed, we have `logs = List.concat [ unseenLogs, seenLogs ]` and not the reverse since logs are stored in a List, by adding new logs at the head of the list for performance reasons. Otherwise adding a new log would be O(n) where n is the list length so getting worse as time passes. That's why there is a `List.reverse` in the view code.
